File: java/integrations/okhttp/src/test/java/com/zourzouvillys/httpsig/okhttp/SigningInterceptorTest.java
package com.zourzouvillys.httpsig.okhttp;

import static org.junit.jupiter.api.Assertions.*;

import java.time.Instant;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.zourzouvillys.httpsig.Algorithm;
import com.zourzouvillys.httpsig.Keys;
import com.zourzouvillys.httpsig.RawMessage;
import com.zourzouvillys.httpsig.SignatureParameters;
import com.zourzouvillys.httpsig.Verifier;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;

class SigningInterceptorTest {

    private MockWebServer server;

    @BeforeEach
    void setUp() throws Exception {
        server = new MockWebServer();
        server.start();
    }

    @AfterEach
    void tearDown() throws Exception {
        server.shutdown();
    }

    @Test
    void interceptorSignsAndVerifierAccepts() throws Exception {
        byte[] secret = "test-secret-long-enough-for-hmac!".getBytes();
        var signingKey = Keys.hmacSHA256Key("test-key", secret);
        var verifyingKey = Keys.hmacSHA256Key("test-key", secret);

        var interceptor = new SigningInterceptor(
            signingKey,
            req -> SignatureParameters.builder()
                .component("@method")
                .component("@authority")
                .component("@path")
                .keyId("test-key")
                .algorithm(Algorithm.HMAC_SHA256)
                .createdEpoch(1618884473L)
                .build()
        );

        var client = new OkHttpClient.Builder()
            .addInterceptor(interceptor)
            .build();

        server.enqueue(new MockResponse().setBody("ok"));

        var request = new Request.Builder()
            .url(server.url("/api/resource"))
            .get()
            .build();

        try (var response = client.newCall(request).execute()) {
            assertEquals(200, response.code());
        }

        RecordedRequest recorded = server.takeRequest();

        // the recorded request should have signature headers
        String sigInput = recorded.getHeader("Signature-Input");
        String sig = recorded.getHeader("Signature");
        assertNotNull(sigInput, "Signature-Input header should be present");
        assertNotNull(sig, "Signature header should be present");

        // reconstruct as RawMessage and verify the signature
        var uri = recorded.getRequestUrl().uri();
        var rawMsg = RawMessage.request(
            recorded.getMethod(),
            uri,
            Map.of(
                "signature-input", List.of(sigInput),
                "signature", List.of(sig)
            )
        );

        var result = Verifier.verify(rawMsg, (keyId, alg) -> verifyingKey,
            Verifier.VerifyOptions.defaults(), null);

        assertEquals("sig1", result.label());
        assertEquals("test-key", result.keyId());
        assertEquals(Algorithm.HMAC_SHA256, result.algorithm());
    }

    @Test
    void customLabelIsUsed() throws Exception {
        byte[] secret = "another-secret-long-enough!!!!!!".getBytes();
        var signingKey = Keys.hmacSHA256Key("k2", secret);
        var verifyingKey = Keys.hmacSHA256Key("k2", secret);

        var interceptor = new SigningInterceptor(
            signingKey,
            "my-sig",
            req -> SignatureParameters.builder()
                .component("@method")
                .keyId("k2")
                .algorithm(Algorithm.HMAC_SHA256)
                .createdEpoch(100L)
                .build()
        );

        var client = new OkHttpClient.Builder()
            .addInterceptor(interceptor)
            .build();

        server.enqueue(new MockResponse().setBody("ok"));

        var request = new Request.Builder()
            .url(server.url("/"))
            .get()
            .build();

        try (var response = client.newCall(request).execute()) {
            assertEquals(200, response.code());
        }

        RecordedRequest recorded = server.takeRequest();
        String sigInput = recorded.getHeader("Signature-Input");
        assertNotNull(sigInput);
        assertTrue(sigInput.startsWith("my-sig="), "should use custom label, got: " + sigInput);

        // verify it too
        var rawMsg = RawMessage.request(
            recorded.getMethod(),
            recorded.getRequestUrl().uri(),
            Map.of(
                "signature-input", List.of(sigInput),
                "signature", List.of(recorded.getHeader("Signature"))
            )
        );

        var opts = new Verifier.VerifyOptions(null, null, null, null, "my-sig", null);
        var result = Verifier.verify(rawMsg, (keyId, alg) -> verifyingKey, opts, null);
        assertEquals("my-sig", result.label());
    }
}

File: kotlin/integrations/okhttp/src/test/kotlin/io/zrz/httpsig/okhttp/SigningInterceptorTest.kt
package io.zrz.httpsig.okhttp

import io.zrz.httpsig.Keys
import io.zrz.httpsig.SignatureParameters
import io.zrz.httpsig.Verifier
import io.zrz.httpsig.RawMessage
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.net.URI

class SigningInterceptorTest {

    private lateinit var server: MockWebServer
    private val secret = "super-secret-key-that-is-at-least-256-bits-long!!".toByteArray()
    private val key = Keys.hmacSHA256Key("test-key", secret)

    @BeforeEach
    fun setUp() {
        server = MockWebServer()
        server.start()
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    @Test
    fun `interceptor adds signature headers to request`() {
        server.enqueue(MockResponse().setResponseCode(200))

        val client = OkHttpClient.Builder()
            .addInterceptor(
                SigningInterceptor(key) { _ ->
                    SignatureParameters.builder()
                        .component("@method")
                        .component("@authority")
                        .keyId("test-key")
                        .createdEpoch(1618884473)
                        .build()
                }
            )
            .build()

        val request = Request.Builder()
            .url(server.url("/test"))
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            assertEquals(200, response.code)
        }

        val recorded = server.takeRequest()
        assertNotNull(recorded.getHeader("Signature-Input"))
        assertNotNull(recorded.getHeader("Signature"))
        assertTrue(recorded.getHeader("Signature-Input")!!.startsWith("sig1="))
        assertTrue(recorded.getHeader("Signature")!!.startsWith("sig1="))
    }

    @Test
    fun `signed request can be verified`() {
        server.enqueue(MockResponse().setResponseCode(200))

        val client = OkHttpClient.Builder()
            .addInterceptor(
                SigningInterceptor(key, "hmac-sig") { _ ->
                    SignatureParameters.builder()
                        .component("@method")
                        .component("@authority")
                        .component("@path")
                        .keyId("test-key")
                        .createdEpoch(1618884473)
                        .build()
                }
            )
            .build()

        val request = Request.Builder()
            .url(server.url("/verify-me"))
            .get()
            .build()

        client.newCall(request).execute().close()

        val recorded = server.takeRequest()
        val sigInput = recorded.getHeader("Signature-Input")!!
        val sig = recorded.getHeader("Signature")!!

        // reconstruct the message on the "server side" and verify the signature
        val serverMsg = RawMessage.request(
            recorded.method!!,
            URI.create(recorded.requestUrl.toString()),
            mapOf(
                "signature-input" to listOf(sigInput),
                "signature" to listOf(sig),
            ),
        )

        val result = Verifier.verify(serverMsg, { _, _ -> key })
        assertEquals("hmac-sig", result.label)
        assertEquals("test-key", result.keyId)
    }

    @Test
    fun `custom label is used in headers`() {
        server.enqueue(MockResponse().setResponseCode(200))

        val client = OkHttpClient.Builder()
            .addInterceptor(
                SigningInterceptor(key, "my-label") { _ ->
                    SignatureParameters.builder()
                        .component("@method")
                        .keyId("test-key")
                        .createdEpoch(1618884473)
                        .build()
                }
            )
            .build()

        val request = Request.Builder()
            .url(server.url("/label-test"))
            .get()
            .build()

        client.newCall(request).execute().close()

        val recorded = server.takeRequest()
        assertTrue(recorded.getHeader("Signature-Input")!!.startsWith("my-label="))
        assertTrue(recorded.getHeader("Signature")!!.startsWith("my-label="))
    }

    @Test
    fun `paramsFactory receives the original request`() {
        server.enqueue(MockResponse().setResponseCode(200))

        var capturedMethod: String? = null
        var capturedPath: String? = null

        val client = OkHttpClient.Builder()
            .addInterceptor(
                SigningInterceptor(key) { req ->
                    capturedMethod = req.method
                    capturedPath = req.url.encodedPath
                    SignatureParameters.builder()
                        .component("@method")
                        .keyId("test-key")
                        .createdEpoch(1618884473)
                        .build()
                }
            )
            .build()

        val request = Request.Builder()
            .url(server.url("/factory-test"))
            .get()
            .build()

        client.newCall(request).execute().close()

        assertEquals("GET", capturedMethod)
        assertEquals("/factory-test", capturedPath)
    }
}
